hotui_server/lib/hotui_server.dart

```diff
@@ -0,0 +1,256 @@
+    );
+  setVMEnvironmentDefines(<String, String>{}, compilerOptions);
+
+  // Create incremental compiler.
+  final Uri initializeFromDillUri =
+      Uri.file(options['initialize-from-dill'] ?? options['output-dill']);
+  final IncrementalCompiler incrementalCompiler = IncrementalCompiler(
+    compilerOptions,
+    entrypointUri,
+    initializeFromDillUri: initializeFromDillUri,
+  );
+
+  // Compile the initial component
+  final Component component =
+      await incrementalCompiler.compile(entryPoint: entrypointUri);
+  component.computeCanonicalNames();
+  incrementalCompiler.accept();
+
+  // Ready for hot UI.
+  HotUIService(
+    incrementalCompiler,
+    component,
+    File(initializeFromDillUri.toFilePath()).parent,
+  );
+}
+
+/// This service avoids re-parsing dart code by maintaining a single full
+/// component and performing updates to this component in place.
+///
+/// Hot ui works by specifiny a libraryId, classId, methodId, and new method
+/// body. The hot UI service uses this information to produce a partial kernel
+/// file containing the updated code. Notably, this is done without modifying
+/// source on disk using an inplace edit of the kernel AST.
+///
+/// Since the hot ui service may accept any libraryId, it must hold a full
+/// component in memory. Then, after modifying the AST, it manually trims the
+/// component to contain only the modified library.
+class HotUIService {
+  HotUIService(this.incrementalCompiler, this.component, this.outputDirectory) {
+    stdin
+        .transform(utf8.decoder)
+        .transform(const LineSplitter())
+        .listen(onMessage);
+    print(kReadyMessage);
+  }
+
+  final IncrementalCompiler incrementalCompiler;
+  final Component component;
+  final Directory outputDirectory;
+
+  // Example invocation:
+  //
+  //    {"class":"_DemoItem","method":"build","library":"package:flutter_gallery/gallery/home.dart","methodBody":"Text('Hello, World')"}
+  Future<void> reloadMethod(String libraryId, String classId, String methodId,
+      String methodBody) async {
+    final Procedure procedure = await incrementalCompiler.compileExpression(
+        methodBody,
+        // This is not yet flexible for non-build expressions.
+        <String>['context'],
+        <String>['BuildContext'],
+        libraryId,
+        classId,
+        false);
+    component.transformChildren(
+        BodyReplacementTransformer(libraryId, classId, methodId, procedure));
+    final Library modifiedLibrary = component.libraries.firstWhere(
+        (Library library) => library.importUri.toString() == libraryId,
+        orElse: () => null);
+    if (modifiedLibrary == null) {
+      throw Exception('Could not find library with id: $libraryId');
+    }
+    final Component partialComponent = Component(libraries: <Library>[
+      modifiedLibrary,
+    ]);
+    partialComponent.unbindCanonicalNames();
+    partialComponent.computeCanonicalNames();
+    final IOSink sink =
```

Comment:
Maybe take the sink as a parameter to reloadMethod so you can just pass in the sink for the http response later on.